In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn import linear_model
from sklearn.datasets import load_breast_cancer,load_boston
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error,mean_absolute_error,accuracy_score
from sklearn.tree import DecisionTreeClassifier,DecisionTreeRegressor
from sklearn import tree
import warnings
warnings.filterwarnings("ignore")

# Görselleştirmeleri görüntülemek için Image fonksiyonunu kulanalım:
from IPython.display import Image

# Soru 1:
Using the data sets you use in the second project, the regression project and the third project, the classification project, create regression and classification models with decision trees and determine the most appropriate parameter values for these models. You can use the cross-validation methods you learned in the previous section to find the ideal parameter.

#### Firstly, load the regression tree

In [13]:
house=load_boston()
house

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

In [20]:
df=pd.DataFrame(house.data, columns=house.feature_names)
df["target"]=house.target
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  target  
0     15.3  396.90   4.98    24.0  
1     17.8  396.90   9.14    21.6  
2     17.8  392.83   4.03    34.7  
3     18.7  394.63   2.94    33.4  
4     18.7  396.90   5.33    36.2

In [21]:
Y=df["target"]
X=df.iloc[:,:13]

In [59]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.20, random_state=42)

reg_tree=DecisionTreeRegressor(criterion="mse", 
                               max_depth=4,
                              max_features=1)

reg_tree.fit(X_train,Y_train)

DecisionTreeRegressor(max_depth=4, max_features=1)

In [60]:
mse=mean_squared_error(y_pred_tree,Y_test)
rmse=np.sqrt(mse)

print("Accuracy of the regression tree : {:.4f}". format(rmse))

Accuracy of the regression tree : 6.9991


#### Most appropriate parameters for the regression tree

In [104]:
param_reg={"criterion":["mse","friedman_mse","mae"],
           "max_depth":np.arange(1,7,1),
           "max_features":np.arange(1,7,1)
          }

grid_reg_tree=GridSearchCV(estimator=reg_tree,
                          param_grid=param_reg,
                          cv=10)

grid_reg_tree.fit(X,Y)
print("Best parameters for regression tree:", grid_reg_tree.best_params_)
print("Best score for regression tree: ", grid_reg_tree.best_score_)

Best parameters for regression tree: {'criterion': 'mse', 'max_depth': 6, 'max_features': 6}
Best score for regression tree:  0.8023484370325225


#### Now, load the classification tree

In [33]:
from sklearn.datasets import load_breast_cancer

In [36]:
cancer=load_breast_cancer()
cancer

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [37]:
df_cancer=pd.DataFrame(cancer.data, columns=cancer.feature_names)
df_cancer["target"]=cancer.target
df_cancer.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [42]:
Y_c=df_cancer["target"]
X_c=df_cancer.iloc[:,:30]

In [43]:
X_c_train, X_c_test, Y_c_train, Y_c_test = train_test_split(X_c, Y_c, test_size=0.20, random_state=42)

In [96]:
class_tree = DecisionTreeClassifier(criterion="entropy",
                                 max_features=2,
                                 max_depth=5
                                 )

class_tree.fit(X_c_train,Y_c_train)

print("Accuracy of classification tree: {:.4f}". format(class_tree.score(X_c_test,Y_c_test)))

Accuracy of classification tree: 0.9649


#### Most appropriate parameters for classification tree:

In [97]:
param_class={"criterion":["entropy","gini"],
             "max_features":np.arange(1,5,1),
             "max_depth":np.arange(1,8,1),
             "min_samples_leaf":np.arange(1,6,1)
            }

grid_class_tree = GridSearchCV(estimator=class_tree,
                               param_grid=param_class,
                               cv=10)

grid_class_tree.fit(X_c_train,Y_c_train)
print("Best parameters of classification tree: ", grid_class_tree.best_params_)
print("Best score of classification tree: ", grid_class_tree.best_score_)

Best parameters of classification tree:  {'criterion': 'gini', 'max_depth': 5, 'max_features': 4, 'min_samples_leaf': 3}
Best score of classification tree:  0.9406280193236716


# Soru 2:
Using below given table answer questions a and b.

* a) Using the dataset above, we want to build a decision tree which classifies Y as T/F given the binary variables A,B,C. Draw the tree that would be learned by the greedy algorithm with zero training error. You do not need to show any computation.
* b) Is this tree optimal (i.e. does it get zero training error with minimal depth)? Explain in less than two sentences. If it is not optimal, draw the optimal tree as well.

 label_dict = {
               'A': ['F','T','T','T'],
               'B': ['F','F','T','T'],
               'C': ['F','T','F','T'],
               'Y':['F','T','T','F']}
 pd.DataFrame(label_dict)       

In [2]:
label_dict = {'A': ['F','T','T','T'],
               'B': ['F','F','T','T'],
               'C': ['F','T','F','T'],
               'Y': ['F','T','T','F']}
label_dict=pd.DataFrame(label_dict)   

In [3]:
label_dict=pd.get_dummies(label_dict, drop_first=True)
label_dict

A_T  B_T  C_T  Y_T
0    0    0    0    0
1    1    0    1    1
2    1    1    0    1
3    1    1    1    0

In [4]:
X_label=label_dict.iloc[:,:3]
Y_label=label_dict.iloc[:,3:4]

X_l_train, X_l_test, Y_l_train, Y_l_test= train_test_split(X_label, Y_label, test_size=0.20, random_state=42)

In [5]:
tree_class=DecisionTreeClassifier()

tree_class.fit(X_label,Y_label)
print("Accuracy of the decision tree:", tree_class.score(X_label,Y_label))

Accuracy of the decision tree: 1.0


In [17]:
from sklearn.tree import export_graphviz
import pydotplus
import graphviz
from IPython.display import Image

In [ ]:
tree_data=export_graphviz(tree_class,
                          out_file=None,
                          feature_names=X_label.columns,
                          filled=True
                         )
graph=pydotplus.graphviz.graph_from_dot_data(tree_data)
Image(graph.create_png())


# Soru 4:
How to prevent overfitting in decision tree?

Apply pruning. Tune the following parameters and re-observe the performance please.
* (1) max_depth: represents how deep your tree will be (1 to 32). More you increase the number, more will be the number of splits and the possibility of overfitting.
* (2) min_samples_split: represents minimum number of samples required to split an internal node. If you use 100% of the samples at each node, it will show underfitting, if you use 10%, it might show overfitting. Tune it accordingly.
* (3) min_samples_leaf: represents min. no. of samples required to be in the leaf node. The more you increase the number, more is the possibility of overfitting.

Irrespective of whatever classifier you are using, apply cross validation. Partition the training data set into 65%-70% training and 30%-35% validation to avoid overfitting. If still the problem of overfitting persists, use Random Forest classifier instead of a single decision tree. RF tackles this problem automatically by reducing variance but without increasing the bias.